ui/ui-components/pages/model/[modelId]/source/[sourceId]/edit.tsx

```diff
@@ -72,27 +115,95 @@ export default function Page(props) {
     }
   }
 
-  const onSubmit = async (event) => {
+  const onSubmit: SubmitHandler<FormData> = async (data, event) => {
     event.preventDefault();
     setLoading(true);
-    const state = source.connection.skipSourceMapping
-      ? "ready"
-      : source.previewAvailable
-      ? undefined
-      : "ready";
 
-    const response: Actions.SourceEdit = await execApi(
+    const isBootstrappingUniqueProperty =
+      data.bootstrap?.propertyKey && data.bootstrap?.propertyType;
+    let bootstrapSuccess = false;
+
+    // Unique Property is being created and need to bootstrap?
+    if (isBootstrappingUniqueProperty) {
+      const payload = {
+        key: data.bootstrap.propertyKey,
+        type: data.bootstrap.propertyType,
+        mappedColumn: data.mapping.sourceColumn,
+      };
+
+      const bootstrapResponse: Actions.SourceBootstrapUniqueProperty =
+        await execApi(
+          "post",
+          `/source/${source.id}/bootstrapUniqueProperty`,
+          payload
+        );
+
+      if (bootstrapResponse?.property) {
+        bootstrapSuccess = true;
+        const prrResponse = await execApi<Actions.PropertiesList>(
+          "get",
+          `/properties`,
+          {
+            includeExamples: true,
+            unique: true,
+            state: "ready",
+            modelId: source.modelId,
+          }
+        );
+
+        if (prrResponse?.properties) {
+          setProperties(prrResponse.properties);
+          setPropertyExamples(prrResponse.examples);
+        }
+      } else if (!Object.keys(bootstrapResponse).length) {
+        errorHandler.set({
+          message: "Unable to create mapping.",
+        });
+        setLoading(false);
+        return;
+      }
+    }
+
+    const mapping = bootstrapSuccess
+      ? {
+          [data.mapping.sourceColumn]: data.bootstrap.propertyKey,
+        }
+      : data.mapping?.propertyKey
+      ? {
+          [data.mapping.sourceColumn]: data.mapping.propertyKey,
+        }
+      : source.mapping;
+
+    const state =
+      source.connection.skipSourceMapping ||
+      data.mapping?.propertyKey ||
+      bootstrapSuccess
+        ? "ready"
+        : source.previewAvailable
+        ? undefined
+        : "ready";
+
+    const response = await execApi<Actions.SourceEdit>(
       "put",
       `/source/${sourceId}`,
-      Object.assign({}, source, { state })
+      { ...source, state, mapping }
     );
+
     if (response?.source) {
       setSource(response.source);
       sourceHandler.set(response.source);
-      if (response.source.state !== "ready") {
+
+      // this source can have a schedule, and we have no schedules yet
+      if (scheduleCount === 0 && response.source.scheduleAvailable) {
```

Comment:
Creating a schedule automatically and redirecting to the schedule comes directly from the mapping page logic so I moved this over here. However, this only works for the first schedule on the page as the `scheduleCount` is pulled from the schedules in the model and not particularly for this source. Should this logic be the same?

Reply:
Yeah - I think the logic should be the same.  We know you want a schedule on your primary source... less likely on the secondary sources